### ISSUE TITLE
Inferring JSX handler types

### ISSUE THREAD

**Issue by silviogutierrez** [Microsoft/TypeScript#7053]
Hi there,

Typescript so far is fantastic. Its JSX support has made full static typing on the front end a reality. I had a suggestion, though maybe I'm missing something obvious. If I implement a JSK template with `noImplicitAny: true`, like so:

```jsx
handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    e.someMethodThatDoesntExist();
};

render() {
    return (
    <form className="LoginForm columns small-6 align-middle" onSubmit={this.handleSubmit}>
    </form>
    );
}
```

The compiler will rightly complain that the second method in `handleSubmit` does not exist.

If, however, I leave out `React.FormEvent` as an explicit type, then the compiler will complain that `e` has an implicit any type. Yet, if I open up the react definition, I see that `onSubmit` is specified and *does* have a typing for the event. Can't this be inferred then?

Finally, if I turn *off* `noImplicitAny`, and type the following (no typing for the event):

```jsx
handleSubmit = (e) => {
    e.preventDefault();
    e.someMethodThatDoesntExist();
};

render() {
    return (
    <form className="LoginForm columns small-6 align-middle" onSubmit={this.handleSubmit}>
    </form>
    );
}
```

The compiler won't complain at all! It would be great if inferring could be done, both for us cautious types using `noImplicitAny: true`, and for those living on the edge with `noImplicitAny` set to false.

Please note this all applies if I use alternate constructs like: `onSubmit={e => this.handleSubmit(e)}` with and without fat arrows for the `handleSubmit` implementation.

Again, thanks for some amazing work.

-
Silvio

**Comment by RyanCavanaugh** [comment]
TypeScript doesn't do whole-program type inference.

To figure out the type of `e` in the second snippet, we'd have to examine your *entire* program to find uses of `handleSubmit`, which in turn would require us to find all uses of the containing class, which in turn would require us to look *again* at your entire program, and so on and so on. We make a (performance + can be reasoned about by humans-ness) / convenience trade-off here by only looking in a finite set of places to infer the type of an expression.

**Comment by silviogutierrez** [comment]
Thanks for the quick response! Avoiding full program inference definitely sounds like it would impact performance.

But in this case, its within the *same* file. I guess I should have stated that this is all within a React component class.

I'm less concerned about my own use case, since I use `noImplicitAny`. The inferring would be nice, but at the very least the compiler tells me it can't be inferred.

I'm more wary of people with implicit anys not having this pointed out, when within the same file, handleSubmit would allow them to call `e.anythingThatDoesntExist()` and think it is safe.

Basically, in non JSX land, could we infer within the same file?

```typescript
// All in the same file foo.ts

function handleSubmit(foo) { // inferred to be of type number[]
    console.log(foo[0]); // no error
    console.log(foo.legnth); // error because of typo.
}

interface Bar {
    callback: (foo: number[]) => void;
}

let spam: Bar = {
    callback: handleSubmit,
} // ERROR!
```

If there's any additional research you need me to do, I'd be happy to.

Thanks again.

Silvio

**Comment by RyanCavanaugh** [comment]
> But in this case, its within the same file

This doesn't particularly help (files can be tens of thousands of lines long), and it's still not a great plan based on our experience with languages that do do this.

The real problem you find if you have a language that does this is that it's *very* difficult to issue useful error messages. Flow (flowtype.org) does this and in any non-trivial program it becomes quickly apparent that you have to write type annotations if you want helpful errors. For example, in this program:

![image](https://cloud.githubusercontent.com/assets/6685088/13015057/d13ca4ce-d16b-11e5-8315-faf1a8b464ff.png)

The error is
![image](https://cloud.githubusercontent.com/assets/6685088/13015062/e59d29a2-d16b-11e5-8698-0508f8fb87e1.png)

But the error isn't on line 2, it's on line 12. Now imagine there are thousands of lines of code in this program instead of 12. How do I figure out that it's line 12 that's causing the problem? Even auditing every invocation of `handleThing` isn't sufficient -- I have to audit every caller of every caller of `handleThing`, or maybe every caller of every caller of every caller of `handleThing`, and so on.

**Comment by silviogutierrez** [comment]
Great explanation. I guess under the React mentality of tiny files, I didn't think of the use case of a massive file. Thanks again.
